### ISSUE TITLE
Fontconfig aliases not respected

### ISSUE THREAD

**Issue by loops** [kovidgoyal/kitty#1085]
Just want to start out by saying thank you for Kitty!  It is the best thing to happen to terminal emulation in a long time.  It's damn near perfect.

One thing that I struggled with a bit when setting up my config is that local fontconfig aliases are not respected.  So for instance if you have something like : 

```
<?xml version='1.0'?>
<!DOCTYPE fontconfig SYSTEM 'fonts.dtd'>
<fontconfig>
  <match>
    <test name="family"><string>Pragmata</string></test>
    <edit name="family" mode="assign" binding="strong">
      <string>PragmataPro Mono Liga</string>
    </edit>
  </match>
</fontconfig>
```

In your fonts.conf,  Kitty will not work if you specify the font family as just "Pragmata"

It seems kitty does all its own font mapping on family name and does not use the fontconfig supplied functions that respect such user configuration.   If the Kitty supplied functions fail to make a match then the "monospace" family name is used instead.  No attempt is made to resolve the family name through fontconfig pattern matching.

It would be much nicer if a way was found to respect local preferences and substitutions.

**Comment by loops** [comment]
```

diff --git a/kitty/fontconfig.c b/kitty/fontconfig.c
index 9123de46..9e3834ca 100644
--- a/kitty/fontconfig.c
+++ b/kitty/fontconfig.c
@@ -151,10 +151,12 @@ fc_match(PyObject UNUSED *self, PyObject *args) {
     PyObject *ans = NULL;
 
     if (!PyArg_ParseTuple(args, "|zpppdd", &family, &bold, &italic, &allow_bitmapped_fonts, &size_in_pts, &dpi)) return NULL;
-    pat = FcPatternCreate();
+    if (family && strlen(family) > 0)
+      pat = FcNameParse((const FcChar8*)family);
+    else
+      pat = FcPatternCreate();
     if (pat == NULL) return PyErr_NoMemory();
 
-    if (family && strlen(family) > 0) AP(FcPatternAddString, FC_FAMILY, (const FcChar8*)family, "family");
     if (!allow_bitmapped_fonts) {
         AP(FcPatternAddBool, FC_OUTLINE, true, "outline");
         AP(FcPatternAddBool, FC_SCALABLE, true, "scalable");
diff --git a/kitty/fonts/fontconfig.py b/kitty/fonts/fontconfig.py
index f10d2f35..2979376c 100644
--- a/kitty/fonts/fontconfig.py
+++ b/kitty/fonts/fontconfig.py
@@ -61,8 +61,9 @@ def find_best_match(family, bold=False, italic=False, monospaced=True):
             candidates.sort(key=score)
             return candidates[0]
 
-    # Use fc-match with a generic family
-    family = 'monospace' if monospaced else 'sans-serif'
+    # Use fc-match instead
+    if monospaced:
+        family += ':mono'
     return fc_match(family, bold, italic)
 
```

This has been working fine for me here, but maybe it won't work for others.  Not sure.

**Comment by kovidgoyal** [comment]
There are many reasons for not using fontconfig.

1) It is not a very good library. Poorly documented, hard to configure, not portable (in the sense of not being suitable for inclusion in application bundles, you pretty much have to use the system fontconfig, no matter how outdated it might be, because the on disk locations/data formats are not stable). 

2) I dont know of a way to do matching using a scoring function. For kitty, I need to be able to take a family name and bold/italic and return a matching set of fonts using a scoring function see find_best_match() in fontconfig.py

3) IIRC it is pretty slow. I dont recall the numbers of the top off my head, but they did not fill me with confidence in fontconfig. To be fair, this likely does not actually matter for kitty performance (beyond startup time) as kitty does not call fontconfig very much.

As you can see, I'm not a fan of fontconfig :)

Now coming to your issue, it is another perfect example of the poor design of fontconfig. It should really be possible to define an alias for a font name. Instead what you have is a way to override the behavior of fc-match, and fc-match alone. So one cannot for example define aliases that work with fc-list and fc-match. So to fix your issue properly, if no match is found with the exact name, call fc_match with the passed in family, the iterate over returned fonts and get the set of distinct family names from those fonts. Then intersect those with the entries in font_map, and then do the scoring as before.

**Comment by kovidgoyal** [comment]
See https://github.com/kovidgoyal/kitty/commit/fefe6c90240c7c7e0005dd2ce5bbf2f616f94052 for exactly why I decided to use fc-list rather fc-match

**Comment by loops** [comment]
Hi Kovid,

Thanks for committing https://github.com/kovidgoyal/kitty/commit/a6949df72789e3c5cdce53d764433fe67a0866e8, it worked to solve the problem I was having.  Just wanted to mention that there is a way to solve the issue you saw in https://github.com/kovidgoyal/kitty/commit/fefe6c90240c7c7e0005dd2ce5bbf2f616f94052 with a proportional font being returned when the family name didn't match : 

 
`diff --git a/kitty/fontconfig.c b/kitty/fontconfig.c
index 9123de46..7ecc03fd 100644
--- a/kitty/fontconfig.c
+++ b/kitty/fontconfig.c
@@ -145,16 +145,20 @@ end:
 static PyObject*
 fc_match(PyObject UNUSED *self, PyObject *args) {
     char *family = NULL;
-    int bold = 0, italic = 0, allow_bitmapped_fonts = 0;
+    int bold = 0, italic = 0, mono = 0, allow_bitmapped_fonts = 0;
     double size_in_pts = 0, dpi = 0;
     FcPattern *pat = NULL;
     PyObject *ans = NULL;
 
-    if (!PyArg_ParseTuple(args, "|zpppdd", &family, &bold, &italic, &allow_bitmapped_fonts, &size_in_pts, &dpi)) return NULL;
+    if (!PyArg_ParseTuple(args, "|zpppdd", &family, &bold, &italic, &mono, &allow_bitmapped_fonts, &size_in_pts, &dpi)) return NULL;
     pat = FcPatternCreate();
     if (pat == NULL) return PyErr_NoMemory();
 
     if (family && strlen(family) > 0) AP(FcPatternAddString, FC_FAMILY, (const FcChar8*)family, "family");
+    if (mono) {
+      AP(FcPatternAddString, FC_FAMILY, (const FcChar8*)"monospace", "family");
+      AP(FcPatternAddInteger, FC_SPACING, FC_MONO, "spacing");
+    }
     if (!allow_bitmapped_fonts) {
         AP(FcPatternAddBool, FC_OUTLINE, true, "outline");
         AP(FcPatternAddBool, FC_SCALABLE, true, "scalable");
@@ -163,6 +167,7 @@ fc_match(PyObject UNUSED *self, PyObject *args) {
     if (dpi > 0) { AP(FcPatternAddDouble, FC_DPI, dpi, "dpi"); }
     if (bold) { AP(FcPatternAddInteger, FC_WEIGHT, FC_WEIGHT_BOLD, "weight"); }
     if (italic) { AP(FcPatternAddInteger, FC_SLANT, FC_SLANT_ITALIC, "slant"); }
+    //printf("fc_match = %s\n", FcNameUnparse(pat));
     ans = _fc_match(pat);
 
 end:
diff --git a/kitty/fonts/fontconfig.py b/kitty/fonts/fontconfig.py
index f10d2f35..34a60da7 100644
--- a/kitty/fonts/fontconfig.py
+++ b/kitty/fonts/fontconfig.py
@@ -61,9 +61,7 @@ def find_best_match(family, bold=False, italic=False, monospaced=True):
             candidates.sort(key=score)
             return candidates[0]
 
-    # Use fc-match with a generic family
-    family = 'monospace' if monospaced else 'sans-serif'
-    return fc_match(family, bold, italic)
+    return fc_match(family, bold, italic, monospaced)
 
 
 def resolve_family(f, main_family, bold, italic):
`
Basically what it does is append a backup family name of "monospace" to the match pattern.  The family name can be a comma separated list which will match in order.  Therefore if a bad name is given, monospace will serve as a backup, but wont override the kitty config file if it's set correctly.

So if you set "font_family      XbadNameX" in kitty.conf, the pattern ends up looking like this:

fc_match = XbadNameX,monospace:weight=200:spacing=100:outline=True:scalable=True


The code you committed does work already, so thanks again.   But if you want to simplify things and remove a bunch of code, you can make fontconfig bend to your will without the manual mapping code.

**Comment by kovidgoyal** [comment]
Interesting, thanks. But I think I'll stick with using fc-list first off, as it allows for using a custom scoring algorithm and matches the way font selection works on other platforms.

**Comment by kovidgoyal** [comment]
I have however added your code to the fc_match function, so it is available in the future, if needed.
